client/protocols/analytics.py

```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
             return False
         return True
 
+    @classmethod
+    def read_history(cls):
+        history = {'questions': {}, 'all_attempts': 0}
+        try:
+            with open(cls.ANALYTICS_FILE, 'rb') as fp:
+                history = pickle.load(fp)
+            log.info('Loaded %d history from %s',
+                     len(history), cls.ANALYTICS_FILE)
+        except (IOError, EOFError) as e:
+            log.info('Error reading from ' + cls.ANALYTICS_FILE + \
+                     ', assume no history')
+        return history
+
+    def log_run(self, messages):
+        """Record this run of the autograder to a local file.
+        """
+        history = self.read_history()
+        history['all_attempts'] += 1
+        analytics = messages['analytics']
+        questions = analytics.get('question', [])
+        grading = 'grading' in messages and messages['grading']
+
+        # Attempt to figure out what question is being worked on
+        if not questions and grading:
+            failed = first_failed_test(self.assignment.specified_tests,
+                                       grading)
+            logging.info('First failed test: %s', failed)
+            if failed:
+                questions = [failed]
+            history['question'] = questions
+
+            # Update earlier question correctness status
+            for saved_q, details in history['questions'].items():
+                finished = details['solved']
+                if not finished and saved_q in grading:
+                    score = grading[saved_q]
+                    details['solved'] = is_correct(score)
+        else:
+            history['question'] = questions
+
+        for question in questions:
```

Comment:
Add a comment explaining what this for loop does on a high level, since this for loop is rather involved.